pandas/io/pickle.py

```diff
@@ -42,7 +46,14 @@ def to_pickle(obj, path, compression='infer', protocol=pkl.HIGHEST_PROTOCOL):
     if protocol < 0:
         protocol = pkl.HIGHEST_PROTOCOL
     try:
-        pkl.dump(obj, f, protocol=protocol)
+        if isinstance(f, zipfile.ZipFile):
```

Comment:
I really don't like it - I don't expect that `_get_handle` returns something other than file-like. Creating temp file is unacceptable - some buffered solution can be introduced.Unfortunately support for buffered zip writing is in Python 3.6, but we can backport it and place in: https://github.com/pandas-dev/pandas/blob/master/pandas/compat/

Python 3.6 supports `mode='w'` in `Zipfile.open` (https://github.com/python/cpython/blob/master/Lib/zipfile.py#L1312) so this whole commit for Python 3.6 could look like:
```
if mode == 'wb':
    f = zipfile.open('data.bin', mode='w')
else:
    # find filename and open for read
```

It should be really easy to backport for Python 3,x - backporting for Python 2.7 can be trickier